azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -66,10 +67,18 @@ def _execute_op(self, op):
                 ca_cert=self.ca_cert,
                 x509_cert=self.client_cert,
             )
-            self.transport.on_mqtt_connected_handler = self._on_mqtt_connected
-            self.transport.on_mqtt_connection_failure_handler = self._on_mqtt_connection_failure
-            self.transport.on_mqtt_disconnected_handler = self._on_mqtt_disconnected
-            self.transport.on_mqtt_message_received_handler = self._on_mqtt_message_received
+            self.transport.on_mqtt_connected_handler = CallableWeakMethod(
+                self, "_on_mqtt_connected"
+            )
```

Comment:
Could this be applied as a decorator?

Reply:
No, but that would be nice.  Well, maybe possible, but that would involve some unusual gymnastics.

To briefly explain: Notice the way the method name is a string rather than a function object.  I did it this way because it's easy to make a weak reference to an object, but it's hard to make a weak reference to a method.  I investigated this far enough to dismiss this idea, but not far enough to explain.   

(The lifetime semantics of a bound method are somewhat somewhat unique because of the binding mechanism.  (3.7 doc for weakref.WeakMethod() calls it  "ephemeral")   In 3.4+, there is a weakref.WeakMethod() object that makes it possible to make a weak reference to a method, but this doesn't exist in 2.7.  I've seem WeakMethod implementations for 2.7, but they are old enough and differently-licensed, so i don't want to dig into them.)


---
In reply to: [332250484](https://github.com/Azure/azure-iot-sdk-python/pull/334#discussion_r332250484) [](ancestors = 332250484)

Reply:
That makes sense, I didn't even think of that, but yeah, best to just avoid. Dealing with the differences in method/function binding across Py 2 vs 3 is a pain.

---
In reply to: [332527083](https://github.com/Azure/azure-iot-sdk-python/pull/334#discussion_r332527083) [](ancestors = 332527083,332250484)